ocaml/xenopsd/xc/device.ml

```diff
@@ -92,6 +90,170 @@ module Profile = struct
         Qemu_upstream_compat
 end
 
+(** File-descriptor event monitor implementation for the epoll library *)
+module Monitor = struct
+  let create () = Polly.create ()
+
+  let add m fd = Polly.add m fd Polly.Events.inp
+
+  let remove m fd = Polly.del m fd
+end
+
+(* Helper module to start up services *)
+module Service = struct
+  type t = {
+      name: string
+    ; domid: domid
+    ; exec_path: string
+    ; pid_path: string
+    ; cancel_path: string
+    ; timeout_seconds: float
+    ; args: string list
+    ; execute: path:string -> args:string list -> domid:domid -> unit -> string
+  }
+
+  let alive service =
+    let is_active = Fe_systemctl.is_active ~service in
+    ( if not is_active then
+        let status = Fe_systemctl.show ~service in
+        error
+          "%s: unexpected termination \
+           (Result=%s,ExecMainPID=%d,ExecMainStatus=%d,ActiveState=%s)"
+          service status.result status.exec_main_pid status.exec_main_status
+          status.active_state
+    ) ;
+    is_active
+
+  type watch_trigger = Finished | Cancelled | Empty
+
+  exception ECancelled of Xenops_task.task_handle
+
+  let raise_e = function
+    | ECancelled t ->
+        Xenops_task.raise_cancelled t
+    | e ->
+        raise e
+
+  let start_and_wait_for_readyness ~task ~service =
+    (* create watches for pidfile and task cancellation *)
+    let notifd = Inotify.create () in
+    let pollfd = Monitor.create () in
+    let ready_watch =
+      Inotify.add_watch notifd
+        (Filename.dirname service.pid_path)
+        [Inotify.S_Create]
+    in
+    let cancel_watch =
+      Inotify.add_watch notifd
+        (Filename.dirname service.cancel_path)
+        [Inotify.S_Create]
+    in
+    Monitor.add pollfd notifd ;
+
+    let close_fds () =
+      Monitor.remove pollfd notifd ;
+      Polly.close pollfd ;
+      Inotify.rm_watch notifd ready_watch ;
+      Inotify.rm_watch notifd cancel_watch ;
+      Unix.close notifd
```

Comment:
Should this be guarded by a try and any exception be logged because it would mean that resources leak?